README.md
foMPI implements the one-sided chapter of MPI-3.0 standard. It offers a
C and a Fortran interface.

This library was written by Robert Gerstenberger, Roberto Belli, Maciej Besta and
Torsten Hoefler and is copyrighted by the ETH-Zurich (c) 2014. 
See LICENSE for details on the license.

Building
========

Compiler
--------
Currently the implementation uses inline assembly for intra-node atomics
and SSE for a fast intra-node memory copy, which isn't supported by the
Cray compiler. We used primarily gcc, but also tested the Intel compiler
and other compiler might work as well.

The Fortran interface can be accessed in the Fortran 77 way (include
'fompif.h') or the Fortran 90 way (use fompi). Since the structure of a
module file depends on the compiler and you wish to build foMPI with a
different compiler than your application, please use the following
scheme:
1) load the compiler with whom you intend to compile foMPI
2) make libfompi.a
3) load the compiler with whom you intend to compile your application
4) rm -f fompi_fortran.o module_fompi.o fompi.{compiler-specific file 
   extension for Fortran module files, .mod for gcc}
5) make libfompi.a
At the moment the name mangling of the Fortran wrapper is hard-coded
(single underscore). If you need to change this for your compiler,
please have a look at line 6 in fompi_fortran.c .

Dependencies
------------
The foMPI tar archive comes with two additional libaries: MPITypes and
libtopodisc, which are included. Just type "make" and those libraries and
the foMPI object files will be built, and packed together in the
ar-archive named libfompi.a. Also a little test program will be built.
If you wish to built the libraries with a different compiler, please use
"make mpitypes" or "make libtopodisc".  

foMPI also needs the m4 macro processor.

* m4 is available from http://www.gnu.org/software/m4/ .
* MPITypes: For more information see the homepage 
  (http://www.mcs.anl.gov/mpitypes/) or their paper from the EuroMPI
  2009 conference (R. Ross, "MPITypes: Processing MPI Datatypes Outside
  MPI").
* Libtopodisc was provided by William Gropp
  (http://www.cs.illinois.edu/~wgropp/ ).

Includes/Linking
----------------
Additional includes to build your application with foMPI are:
-I{PATH to foMPI}

Additional parameter for linking your application with foMPI are:
-L{PATH to foMPI} -lfompi
-ldmapp

Build parameter
---------------
The building parameter can be found in the Makefile.inc file.
You can choose your appropriate compiler (CC), flags (CCFLAGS), etc.
Additionally foMPI provides a section in the file fompi_internal.h,
that let you influence some of the inner work of foMPI:
* DXPMEM (line 11): Enables/disables XPMEM support for intra-node
  communication
* DTYPE_HANDLING (line 14): Selects the strategy to handle MPI derived
  datatypes for putting data with inter-node targets. A more detailed
  explanation can be found in the section about MPI derived datatypes.

If you experience problems with foMPI, please try removing the 
XPMEM support.


Using foMPI
===========

All MPI one-sided functions are prefixed with foMPI instead of MPI.
Additional all constants like foMPI_LOCK_{SHARED,EXLCUSIVE}, the error
codes, asserts (though not supported at the moment), flavor, memory
model and MPI_Ops for the accumulate-derived communication functions
need to be prefixed with foMPI. Below you find a complete list of
all implemented constants.

The datatypes MPI_Win and MPI_Request (if used with foMPI RMA functions
that take requests) need to be renamed to foMPI_Win or foMPI_Request.
foMPI_Requests are not compatible with MPI_Requests, and can only be
used with foMPI wait and test functions.

Some functions needed to be overloaded to have foMPI work properly
(MPI_Init, MPI_Finalize, MPI_Type_free). The wait and test functions for
foMPI requests can't be mixed with normal MPI requests. For a full list
of overloaded functions, please check the list below.

Apart from prefixing elements of the MPI one-sided scope the function
interfaces are declared as in the standard. See fompi.h for the full
function prototypes.

Below you can find a complete list of implemented and not implemented
functions of the one-sided chapter.

foMPI implements the unified memory model.

MPI Datatypes
-------------

foMPI supports all datatypes that are supported by MPITypes for
foMPI_Put/foMPI_Get. The following datatypes are in the fast path for
these two communication functions:
* MPI_CHAR
* MPI_UINT64_T
* MPI_DOUBLE
* MPI_INT

All MPI datatypes are supported for the accumulate-based communication
functions except:
* handles from MPI_TYPE_CREATE_F90_INTEGER
* MPI_INTEGER16
* handles from MPI_TYPE_CREATE_F90_REAL
* MPI_REAL2
* MPI_CXX_BOOL
* handles from MPI_TYPE_CREATE_F90_COMPLEX
* MPI_CXX_FLOAT_COMPLEX
* MPI_CXX_DOUBLE_COMPLEX
* MPI_CXX_LONG_DOUBLE_COMPLEX

The following combinations of MPI datatypes and MPI_Ops are
hardware-supported:
* MPI_DOUBLE/MPI_{BAND/BOR/BXOR}
* MPI_INT64_T/MPI_{BAND/BOR/BXOR}
* MPI_INT64_T/MPI_SUM

foMPI supports two different modes to handle MPI derived datatypes for
putting data with inter-node targets:
* The maximal block strategy puts each block of the superposition of the
  two datatypes separately. This was the original bufferless approach
  presented in the foMPI SC'13 paper and was the default strategy until
  version 0.2.2.
* The fixed buffer size strategy copies elements into a temporary buffer
  until a predefined size (= fixed size) is reached, and then the
  temporary buffer is put. This strategy allows the overlapping of
  copying and communication. This strategy is the current default
  (beginning with version 0.2.2).
The strategy can be selected in fompi_internal.h, line 14 with the
DTYPE_HANDLING constant. If the fixed buffer size strategy is selected,
additional parameter are active:
* DTYPE_FIXED_SIZE = size of one buffer = maximum size of a underlying
  DMAPP put operation
* DTYPE_BUFFER_NUMBER = number of buffer that can be used in parallel
The maximal block strategy is always used for intra-node communication.

For a detailed theoretical analysis of both strategies see:
----------------------------------------------------------------------
T. Schneider, R. Gerstenberger, T. Hoefler: Compiler Optimizations for
Non-Contiguous Remote Data Movement. In: Proceedings of 26th
International Workshop on Languages and Compilers for Parallel
Computing (LCPC'13)
----------------------------------------------------------------------

For details on the MPI Datatypes implementation in foMPI:
----------------------------------------------------------------------
R. Gerstenberger, T. Hoefler: Handling Datatypes in MPI-3 One Sided.
ACM Students Research Competition Poster at the The International
Conference for High Performance Computing, Networking, Storage and
Analysis (SC’13).
http://sc13.supercomputing.org/sites/default/files/PostersArchive/spost114.html
----------------------------------------------------------------------

Fortran
-------
All communication functions take a displacement parameter, which is of
the integer(kind=MPI_ADDRESS_KIND) type. To ensure compatibility with the
C wrapper code, please use a variable for the displacement parameter
instead of using a constant (which will be transfered as 4 byte integer
instead of the bigger integer(kind=MPI_ADDRESS_KIND)).

On our test system, we experienced problems with MPI_Get_address, which
gives different results for the same location if called from C and
Fortran. Since our library is implemented in C, this is an issue if you
intend to use dynamic windows.
We included a test_address subdirectory, which contains a small test
program for this problem. If all three locations are the same, you are
fine. If not, please use foMPI_Get_address instead.


Questions
=========

If you have additional questions, feel free to contact me under the
following e-mail address: Robert Gerstenberger <gerro@illinois.edu>.


Citation
========

Any published work which uses this software should include the 
following citation:

----------------------------------------------------------------------
R. Belli T. Hoefler: Notified Access: Extending Remote Memory Access Programming Models
for Producer-Consumer Synchronization. In: Proceedings
of the 29th IEEE International Parallel & Distributed Processing Symposium
(IPDPS'15)
----------------------------------------------------------------------

----------------------------------------------------------------------
R. Gerstenberger, M. Besta and T. Hoefler: Enabling Highly-Scalable
Remote Memory Access Programming with MPI-3 One Sided. In: Proceedings
of the International Conference on High Performance Computing,
Networking, Storage and Analysis (SC'13)
----------------------------------------------------------------------


Implemented function
====================

Window creation
---------------
* foMPI_Win_create
* foMPI_Win_allocate
* foMPI_Win_create_dynamic
* foMPI_Win_free

* foMPI_Win_attach
* foMPI_Win_detach


Synchronization
---------------
* foMPI_Win_fence

* foMPI_Win_post
* foMPI_Win_start
* foMPI_Win_complete
* foMPI_Win_wait
* foMPI_Win_test

* foMPI_Win_lock
* foMPI_Win_unlock
* foMPI_Win_lock_all
* foMPI_Win_unlock_all

* foMPI_Win_flush
* foMPI_Win_flush_all
* foMPI_Win_flush_local
* foMPI_Win_flush_local_all
* foMPI_Win_sync

* foMPI_Notify_init


Communication
-------------
* foMPI_Put
* foMPI_Get
* foMPI_Accumulate
* foMPI_Get_accumulate
* foMPI_Fetch_and_op
* foMPI_Compare_and_swap

* foMPI_Rput
* foMPI_Rget
* foMPI_Raccumulate
* foMPI_Rget_accumulate

* foMPI_Put_notify
* foMPI_Get_notify

Miscellaneous
-------------
* foMPI_Win_get_group

* foMPI_Win_set_name
* foMPI_Win_get_name


Overloaded functions
====================

Additional to MPI one-sided functions also the following functions need
to be overloaded:

* foMPI_Type_free
* foMPI_Init
* foMPI_Finalize

All of them above will call their MPI equivalent with the passed
parameters, but do additional work for the foMPI implementation.

* foMPI_Wait
* foMPI_Test
* foMPI_Waitall
* foMPI_Waitany
* foMPI_Start
* foMPI_Request_free

These functions can only be used together with foMPI_Requests and can't
handle normal MPI_Request. All other wait oder test routines are not
implemented at the moment.


Not implemented functions
=========================

* foMPI_Win_allocate_shared
* foMPI_Win_shared_query
* foMPI_Win_get_attr
* foMPI_Win_set_attr
* foMPI_Win_delete_attr
* foMPI_Win_set_errhandler
* foMPI_Win_call_errhandler
* foMPI_Win_set_info
* foMPI_Win_get_info


Constants
=========

Create flavor
-------------
* foMPI_WIN_FLAVOR_CREATE
* foMPI_WIN_FLAVOR_ALLOCATE
* foMPI_WIN_FLAVOR_DYNAMIC


Memory model
------------
* foMPI_WIN_SEPARATE
* foMPI_WIN_UNIFIED


Asserts
-------
* foMPI_MODE_NOCHECK
* foMPI_MODE_NOSTORE
* foMPI_MODE_NOPUT
* foMPI_MODE_NOPRECEDE
* foMPI_MODE_NOSUCCEED
Asserts are not used at the moment, but will be accepted by the library.


Accumulate operations
---------------------
* foMPI_SUM
* foMPI_PROD
* foMPI_MAX
* foMPI_MIN
* foMPI_LAND
* foMPI_LOR
* foMPI_LXOR
* foMPI_BAND
* foMPI_BOR
* foMPI_BXOR
* foMPI_MAXLOC
* foMPI_MINLOC
* foMPI_REPLACE
* foMPI_NO_OP


Lock type
---------
* foMPI_LOCK_SHARED
* foMPI_LOCK_EXCLUSIVE


Error codes
-----------
* foMPI_ERROR_RMA_NO_SYNC
* foMPI_ERROR_RMA_DATATYPE_MISMATCH
* foMPI_ERROR_RMA_NO_LOCK_HOLDING
* foMPI_ERROR_RMA_SYNC_CONFLICT
* foMPI_ERROR_RMA_WIN_MEM_NOT_FOUND
* foMPI_OP_NOT_SUPPORTED
* foMPI_DATATYPE_NOT_SUPPORTED
* foMPI_NAME_ABBREVIATED


CHANGES
0.2.3
*added additional functionallity to the fortran warpper
*fixed problem with the node local notification matching
*return number of matched notifications using MPI_Status_set_elements
*fixed gcc5 compatibility problem 
*removed starvation hack

0.2.2
*added UGNI and XPMEM support for RMA notifications.
*main calls introduced: foMPI_Put_notify, foMPI_Get_notify, foMPI_Notify_init

0.2.1
* removed autoconf dependency for libtopodisc to allow an easier built (proposal by Timo Schneider)
* fixed problem with compiling the Fortran foMPI_Win_allocate interface with the Intel compiler
* fixed problem with Fortran window and request book keeping
* extended testing with Intel compiler

0.2.0
* added Fortran interface


fompi_helper.c
// Copyright (c) 2014 ETH-Zurich. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <assert.h>
#include <unistd.h>

#include "fompi.h"

void foMPI_Set_window_mutex( dmapp_return_t* status, dmapp_pe_t target_pe, foMPI_Win_desc_t* win_ptr, dmapp_seg_desc_t* win_ptr_seg, foMPI_Win win ) {

  int64_t result;

 /* TODO: busy waiting */

  do {
	//	do {

	//		*status = dmapp_get( &result, (void*) &(win_ptr->mutex), win_ptr_seg, target_pe, 1, DMAPP_QW);
  //		_check_status(*status, DMAPP_RC_SUCCESS,  (char*) __FILE__, __LINE__);

	//	} while (result != foMPI_MUTEX_NONE);

  //  *status = dmapp_acswap_qw( &result, (void*) &(win_ptr->mutex), win_ptr_seg, target_pe, (int64_t) foMPI_MUTEX_NONE, (int64_t) win->commrank );
    *status = dmapp_acswap_qw( &result, (void*) &(win_ptr->mutex), win_ptr_seg, target_pe, (int64_t) foMPI_MUTEX_NONE, (int64_t) 0 );
    _check_dmapp_status(*status, DMAPP_RC_SUCCESS,  (char*) __FILE__, __LINE__);

			//printf("[%i] kurwa mac\n",win->commrank);
  } while (result != foMPI_MUTEX_NONE);
}

void foMPI_Release_window_mutex( dmapp_return_t* status, dmapp_pe_t target_pe, foMPI_Win_desc_t* win_ptr, dmapp_seg_desc_t* win_ptr_seg, foMPI_Win win ) {

  int64_t result;

  result = foMPI_MUTEX_NONE;
  /* TODO: nonblocking ? */
 // *status = dmapp_put( (void*) &(win_ptr->mutex), win_ptr_seg, target_pe, &result, 1, DMAPP_QW); 
    *status = dmapp_acswap_qw( &result, (void*) &(win_ptr->mutex), win_ptr_seg, target_pe, (int64_t) 0, (int64_t) foMPI_MUTEX_NONE );
  _check_dmapp_status(*status, DMAPP_RC_SUCCESS,  (char*) __FILE__, __LINE__);
}

int get_free_elem( fortran_free_mem_t **list_head ) { /* as a side effect also removes the first element from the list */

  int index;
  fortran_free_mem_t *entry;

  if (*list_head == NULL) {
    return -1; /* no free element left */
  }

  entry = *list_head;
  *list_head = entry->next;
  index = entry->index;

  _foMPI_FREE( entry );

  return index;
}

#ifdef XPMEM
/**
This is executed on the rank who has the buffer that should be exposed at ptr

provided by Timo Schneider, ETH Zurich
*/

/* TODO: static inline */
xpmem_seg_desc_t foMPI_export_memory_xpmem(void *ptr, int len) {

  xpmem_seg_desc_t ret;
  
  void *pagestart = (void*)(((unsigned long)ptr / getpagesize()) * getpagesize());
  ret.offset = (unsigned long) ptr - (unsigned long) pagestart;
  //FIXME might be one page too much
  size_t szpages = ((unsigned long)(ret.offset + len) / getpagesize() + 1) * getpagesize();
  //possible fix: if( ((unsigned long)(ret.offset + len) % getpagesize()) == 0) szpages--;
  ret.seg = xpmem_make(pagestart, szpages, XPMEM_PERMIT_MODE, (void *)0666);
  assert( ret.seg != -1 );

  return ret;
}

/**
This is called by the process that wants to write into a memory region exported
by someone else.

provided by Timo Schneider, ETH Zurich
*/

/* TODO: static inline */
void* foMPI_map_memory_xpmem(xpmem_seg_desc_t rseg, int len, xpmem_apid_t* apid, int* offset) {

  *apid = xpmem_get(rseg.seg, XPMEM_RDWR, XPMEM_PERMIT_MODE, NULL);
  //assert( *apid != -1 );
  if ( *apid == -1 ) {
    printf("%i: apid == -1\n", glob_info.debug_pe);
  }

  *offset = rseg.offset;
  //printf("%i: map offset = %i\n", debug_pe, *offset);

  struct xpmem_addr addr;
  addr.apid = *apid;
  addr.offset = 0;
  int szpages = ((unsigned long)(rseg.offset + len) / getpagesize() + 1) * getpagesize();
  void *map = xpmem_attach(addr, szpages, NULL);
  assert( map != (void*)-1 );
  // correct address in mapping
  return ((void*)((char*)map + rseg.offset));
}

/**
This is called by the process that wanted to free his access to a memory region exported
by someone else.
*/

/* TODO: static inline */
void foMPI_unmap_memory_xpmem( xpmem_apid_t* apid, void* addr, int offset ) {

#ifndef NDEBUG
  int ret;
#endif
 
  //printf("%i: unmap offset = %i\n", debug_pe, offset); 
#ifdef NDEBUG
  xpmem_detach( (char*)addr-offset );
#else  
  ret = xpmem_detach( (char*)addr-offset );
#endif
  assert( ret == 0 );

#ifdef NDEBUG
  xpmem_release( *apid );
#else
  ret = xpmem_release( *apid );
#endif
  assert( ret == 0 );
}

/**
This is called by the process that exported a memory region to unexport this region.
*/

/* TODO: static inline */
void foMPI_unexport_memory_xpmem( xpmem_seg_desc_t rseg ) {

#ifndef NDEBUG
  int ret;
#endif

#ifdef NDEBUG
  xpmem_remove( rseg.seg );
#else
  ret = xpmem_remove( rseg.seg );
#endif
  assert( ret == 0 );
}

int foMPI_onnode_rank_global_to_local( int grank, foMPI_Win win ) {
  int i;

  assert( (grank >= 0 ) && (grank <= win->commsize) );

  if ( ( grank >= win->onnode_lower_bound ) && ( grank <= win->onnode_upper_bound ) ) { /* fast range search */
    return grank-win->onnode_lower_bound;
  } else {
    if ( win->onnode_lower_bound == -1 ) {
      for( i=0 ; i<win->onnode_size ; i++) { /* slow search, TODO: faster implementation */
        if( win->onnode_ranks[i] == grank) {
          return i;
        }
      }
    }
    return -1; /* not a onnode process */
  }
}

int foMPI_onnode_rank_local_to_global( int lrank, foMPI_Win win ) {

  assert( (lrank >= 0) && (lrank < win->onnode_size) );

  if( win->onnode_lower_bound != -1 ) {
    return lrank+win->onnode_lower_bound;
  } else {
    return win->onnode_ranks[lrank];
  }

}
#endif

#ifdef PAPI
#include <string.h>

#include <papi.h>
//! handles all the time measurement

static int counter;

static char epoch_name[23][20];
static int ids[1024];

static MPI_File filehandle_values;

/* PAPI relevant variables */
static int PapiNumberEvents; // max 2
static int PapiEventSet;
static int PapiEventCode[2];

static long long PapiEventCounts[2048];
static char PapiEventName1[20] = "NA\0";
static char PapiEventName2[20] = "NA\0";

//! writes the timing values to the output file
//! if the last parameter is set, then all the unused epochs are also printed

static void timing_print( int last ) {

  char line[256];
  int i;

  char comparand[20];
  snprintf( comparand, 20, "NA" );

  char eventcount1[20];
  char eventcount2[20];

  for( i=0; i<counter ; i++ ) {

// select the right values for PAPI columns
    snprintf( eventcount1, 20, "%lli",  PapiEventCounts[2*i] );
    if (strncmp(PapiEventName1, comparand, 20) == 0 ) { /* handling of event counter that are not set */
      snprintf( eventcount1, 20, "NA");
    }
    snprintf( eventcount2, 20, "%lli",  PapiEventCounts[2*i+1] );
    if (strncmp(PapiEventName2, comparand, 20) == 0 ) { /* handling of event counter that are not set */
      snprintf( eventcount2, 20, "NA");
    }

//! if a epoch id is not in 1..4 is found, only the id is printed (instead of
//! the the epoch name)
    snprintf(line, 256, "%20s%20s%20s%20s%20s\n", epoch_name[ids[i]], PapiEventName1, eventcount1, PapiEventName2, eventcount2 );
    MPI_File_write(filehandle_values, line, strlen(line), MPI_CHAR, MPI_STATUS_IGNORE);
  }

  if (last == 1) {
    long long PapiValue[PapiNumberEvents];
    //stop the PAPI counter
    PAPI_stop( PapiEventSet, PapiValue );
    PAPI_reset( PapiEventSet );
  }

}

//! opens a file handle, to where the timing values are written
static void timing_open_file() {
      
  char filename[20] = "mpi3rma-ex.out\0";
  char line[256];
  int ier;

  ier = MPI_File_open( MPI_COMM_SELF, filename, MPI_MODE_EXCL + MPI_MODE_CREATE + MPI_MODE_WRONLY, MPI_INFO_NULL, &filehandle_values );

  if ( ier != MPI_SUCCESS ) {
    printf("Error at open file %s for writing the timing values. The file probably already exists.\nWill now abort.\n", filename);
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }
  snprintf(line, 256, "%20s%20s%20s%20s%20s\n", "id", "papi_evt1_type", "papi_evt1_val", "papi_evt2_type", "papi_evt2_val");
  MPI_File_write(filehandle_values, line, strlen(line), MPI_CHAR, MPI_STATUS_IGNORE);
}

static void init_papi() {

  int retval;

  retval = PAPI_library_init(PAPI_VER_CURRENT);
  if (retval != PAPI_VER_CURRENT && retval > 0) {
    fprintf(stderr,"PAPI library version mismatch!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }

  if (retval < 0) {
    fprintf(stderr,"PAPI init error!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }

  retval = PAPI_is_initialized();
  if (retval != PAPI_LOW_LEVEL_INITED) {
    fprintf(stderr,"PAPI library not initialized!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }

  PapiEventSet = PAPI_NULL;
  if (PAPI_create_eventset(&PapiEventSet) != PAPI_OK) {
    fprintf(stderr,"PAPI could not create event set!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }

  //get the PAPI events
  char* papi_evt1 = getenv("PAPI_EVT1");
  char* papi_evt2 = getenv("PAPI_EVT2");

  PapiNumberEvents = 0;

  if ( (papi_evt1 != NULL) && (strlen(papi_evt1) != 0)) {
    retval = PAPI_event_name_to_code( papi_evt1, &PapiEventCode[0] );
    if ( retval != PAPI_OK ) {
      switch( retval ) {
        case PAPI_ENOTPRESET:   fprintf(stderr,"The event code provided in the environmental variable PAPI_EVT1 is not a valid PAPI preset!\n"); break;
        case PAPI_ENOEVNT:      fprintf(stderr,"The event provided in the environmental variable PAPI_EVT1 is not available on the underlying hardware!\n"); break;
        default:                fprintf(stderr,"The event code provided in the environmental variable PAPI_EVT1 is not correct!\n"); 
      }
      MPI_Abort( MPI_COMM_WORLD, 1 );
    }
    strncpy( PapiEventName1, papi_evt1, 20 );

    PapiNumberEvents++;
    
    if (PAPI_add_event(PapiEventSet, PapiEventCode[0]) != PAPI_OK) {
      fprintf(stderr,"PAPI could not add counter to event set!\n");
      MPI_Abort( MPI_COMM_WORLD, 1 );
    }
  }

  if ( (papi_evt2 != NULL) && (strlen(papi_evt2) != 0) ) {
    retval = PAPI_event_name_to_code( papi_evt2, &PapiEventCode[1] );
    if ( retval != PAPI_OK ) {
      switch( retval ) {
        case PAPI_ENOTPRESET:   fprintf(stderr,"The event code provided in the environmental variable PAPI_EVT2 is not a valid PAPI preset!\n"); break;
        case PAPI_ENOEVNT:      fprintf(stderr,"The event provided in the environmental variable PAPI_EVT2 is not available on the underlying hardware!\n"); break;
        default:                fprintf(stderr,"The event code provided in the environmental variable PAPI_EVT2 is not correct!\n"); 
      }
      MPI_Abort( MPI_COMM_WORLD, 1 );
    }
    strncpy( PapiEventName2, papi_evt2, 20 );

    PapiNumberEvents++;
  
    if ( PAPI_add_event(PapiEventSet, PapiEventCode[1]) != PAPI_OK) {
      fprintf(stderr,"PAPI could not add counter to event set!\n");
      MPI_Abort( MPI_COMM_WORLD, 1 );
    }
  } 

  if ( PapiNumberEvents == 0 ) {
    fprintf(stderr, "Althrough the mpi3rma code was compiled with PAPI support, the environmental variables PAPI_EVT1 and PAPI_EVT2 are not set.\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }
}

static void cleanup_papi() {
  if ( PAPI_cleanup_eventset(PapiEventSet) != PAPI_OK ) {
    fprintf(stderr,"PAPI could not clean up event set!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }
  if ( PAPI_destroy_eventset(&PapiEventSet) != PAPI_OK ) {
    fprintf(stderr,"PAPI could not destroy event set!\n");
    MPI_Abort( MPI_COMM_WORLD, 1 );
  }
}

//! closes the above mentioned filehandle
static void timing_close_file() {
  MPI_File_close( &filehandle_values );
}
//! sets some init parameter like the testname, method and number of bytes
//! also sets static the epoch names, the epoch_used array and the initial timing with a special id (zero)

void timing_init() {

  counter = 0;

  snprintf( &epoch_name[0][0], 20, "reference");
  snprintf( &epoch_name[1][0], 20, "put_before_dmapp");
  snprintf( &epoch_name[2][0], 20, "put_dmapp");
  snprintf( &epoch_name[3][0], 20, "put_after_dmapp");
  snprintf( &epoch_name[4][0], 20, "flush");
  snprintf( &epoch_name[5][0], 20, "get_before_dmapp");
  snprintf( &epoch_name[6][0], 20, "get_dmapp");
  snprintf( &epoch_name[7][0], 20, "get_after_dmapp");
  snprintf( &epoch_name[8][0], 20, "unlock_before_flush");
  snprintf( &epoch_name[9][0], 20, "unlock_after_flush");
  snprintf( &epoch_name[10][0], 20, "lock");
  snprintf( &epoch_name[11][0], 20, "start");
  snprintf( &epoch_name[12][0], 20, "complete");
  snprintf( &epoch_name[13][0], 20, "post");
  snprintf( &epoch_name[14][0], 20, "wait");
  snprintf( &epoch_name[15][0], 20, "fence");
  snprintf( &epoch_name[16][0], 20, "lock_all");
  snprintf( &epoch_name[17][0], 20, "unlock_all");
  snprintf( &epoch_name[18][0], 20, "cswap");
  snprintf( &epoch_name[19][0], 20, "flush_all");
  snprintf( &epoch_name[20][0], 20, "fetch_and_op");
  snprintf( &epoch_name[21][0], 20, "accumulate");
  snprintf( &epoch_name[22][0], 20, "get_accumulate");

  init_papi();

  if (glob_info.debug_pe == 0) {
    timing_open_file();
  }

  /* start counter */
  PAPI_start( PapiEventSet);
  timing_record( -1 );
  timing_record( 0 );
}

void timing_close() {
  if (glob_info.debug_pe == 0) {
    timing_print( 1 );
  }

  cleanup_papi();

  if (glob_info.debug_pe == 0) {
    timing_close_file();
  }
}

//! records the timing for the given epoch id
//! if number of recorded timings exceed 1024, then the function empties
//! the buffer and they will be printed

void timing_record( int id ) {

  PAPI_stop( PapiEventSet, &PapiEventCounts[2*counter] );
  PAPI_reset( PapiEventSet );

  ids[counter] = id;

  if ( id > -1  ) { 
    if ( ++counter > 1024 ) {
      if (glob_info.debug_pe == 0) {
        timing_print(0);
      }
      counter = 0;
    }
  }
  PAPI_start( PapiEventSet );
}
#endif
